modules/fxgui/src/edu/kit/iti/algover/rule/script/ScriptController.java

```diff
@@ -163,55 +191,124 @@ public void setProof(Proof proof) {
+        if(checkpoint != null) {
+            showSelectedSelector(checkpoint);
+        }
+        //updateInsertPosition();
+        updateInsertPosition(observableInsertPosition.get());
         switchViewedNode();
         view.highlightLine();
     }
 
-    private void showSelectedSelector(ProofNodeCheckpoint checkpoint) {
-        view.getGutterAnnotations().forEach(gutterAnnotation -> gutterAnnotation.setProofNodeIsSelected(false));
-        view.getGutter().getLineAnnotation(checkpoint.caretPosition.getLineNumber()-1).setProofNodeIsSelected(true);
-    }
+    /*private void updateInsertPosition() {
+        Position caretPosition = computePositionFromCharIdx(view.getCaretPosition(), view.getText());
+        ProofNodeCheckpoint insertCheckpoint = getCheckpointForPosition(caretPosition);
+        if(insertCheckpoint != null) {
+            setObservableInsertPosition(new Position(insertCheckpoint.position.getLineNumber() + 1, 0));
+        } else {
+            //no checkpoint found so the script should be empty
+            setObservableInsertPosition((new Position (1, 0)));
+        }
+    }*/
 
+    private void updateInsertPosition(Position oldInsertPos) {
 
-    private void switchViewedNode() {
-        Position caretPosition = getObservableInsertPosition();
+        Position caretPosition = computePositionFromCharIdx(view.getCaretPosition(), view.getText());
+        ProofNodeCheckpoint insertCheckpoint = getCheckpointForPosition(caretPosition);
+        ProofNodeCheckpoint oldInsertCheckpoint = getCheckpointForPosition(oldInsertPos);
+
+        int oldSize = this.sizeOfCheckpoints;
+        int newSize = checkpoints.size();
 
-        ProofNodeCheckpoint checkpoint = getCheckpointForCaretPosition(caretPosition);
+        //no new checkpoints added -> we only moved cursor
+        if(insertCheckpoint != null) {
+            //no new checkpoints added -> we only moved cursor => set observableinsertPos to caretlinenumer+1
+            if(oldSize == newSize) {
+                //the position was probably changed by setting cursor somewhere else
+                setObservableInsertPosition(new Position(insertCheckpoint.position.getLineNumber() + 1, 0));
 
-        // If the selector points to nowhere, its probably because the rule closed the proof and didn't generate
-        // another child...
-        // REVIEW This is kind of ugly. In the future this off-by-one fix has to be removed
-        if (!checkpoint.selector.optionalGet(proof).isPresent()) {
-            if(checkpoint.selector.getParentSelector() != null) {
-                this.listener.onSwitchViewedNode(checkpoint.selector.getParentSelector());
             } else {
-                this.listener.onSwitchViewedNode(checkpoint.selector);
+                //we have new checkpoints and need to find the position where to add the next command
+                int difference = newSize - oldSize;
+                assert difference >= 0;
```

Comment:
when playing around this assertion failed once. Unfurtunatly i cant reproduce this